language/runtime/ceylon/language/Array.java

```diff
@@ -1937,51 +1937,70 @@ else if (step == 1) {
             long sourcePosition){
         return 0L;
     }
 
     @Ignore
-    public long copyTo$length(Array<Element> destination, 
+    public long copyTo$length(Array<? super Element> destination, 
             long sourcePosition, long destinationPosition){
         return Math.min(size-sourcePosition,
                 destination.getSize()-destinationPosition);
     }
 
     @Ignore
-    public void copyTo(Array<Element> destination){
+    public void copyTo(Array<? super Element> destination){
         copyTo(destination, 0L, 0L);
     }
 
     @Ignore
-    public void copyTo(Array<Element> destination, long sourcePosition) {
+    public void copyTo(Array<? super Element> destination, long sourcePosition) {
         copyTo(destination, sourcePosition, 0L);
     }
 
     @Ignore
-    public void copyTo(Array<Element> destination, 
+    public void copyTo(Array<? super Element> destination, 
             long sourcePosition, long destinationPosition) {
         copyTo(destination, sourcePosition, destinationPosition, 
                 copyTo$length(destination, sourcePosition, destinationPosition));
     }
 
-    public void copyTo(@Name("destination") Array<Element> destination, 
+    public void copyTo(@Name("destination")
+                       @TypeInfo("ceylon.language::Array<in Element>") Array<? super Element> destination,
                        @Name("sourcePosition") @Defaulted long sourcePosition, 
                        @Name("destinationPosition") @Defaulted long destinationPosition, 
                        @Name("length") @Defaulted long length) {
         if (length>0) {
             try {
-                arraycopy(array, 
-                        Util.toInt(sourcePosition), 
-                        destination.array, 
-                        Util.toInt(destinationPosition), 
-                        Util.toInt(length));
+                System.out.println("" + elementType + " to " + destination.elementType);
+                if (elementType == destination.elementType) {
+                    arraycopy(array,
+                            Util.toInt(sourcePosition),
+                            destination.array,
+                            Util.toInt(destinationPosition),
+                            Util.toInt(length));
+                }
+                else {
+                    // unbox as necessary
+                    int srcIndex = Util.toInt(sourcePosition);
+                    int dstIndex = Util.toInt(destinationPosition);
+                    int len = Util.toInt(length);
+
+                    if (srcIndex < 0 || dstIndex < 0
+                            || srcIndex + len > getSize()
+                            || dstIndex + len > destination.getSize()) {
+                        throw new AssertionError("Index out of bounds");
+                    }
+                    while (len-- > 0) {
+                        destination.set(dstIndex++, getFromFirst(srcIndex++));
```

Comment:
@jvasileff it might be the case that right here, we know for sure that `destination` is backed by an `Object[]`, and can just use direct assignment to the array element.